circuit/src/opinion/mod.rs

```diff
@@ -279,11 +288,11 @@ where
 					|| "constraint equality",
 					|region: Region<'_, N>| {
 						let mut ctx = RegionCtx::new(region, 0);
-						ctx.assign_advice(
+						let msg_hash = ctx.assign_advice(
 							common.advice[0],
 							Value::known(P::compose(self.msg_hash[i].integer.limbs)),
```

Comment:
This is not enough, we have to compose it using MainGate.

Reply:
You mean we need a new chipset which calculates composed value of the given `AssignedCell` limbs?
We need composed value of the `AssignedPublicKey` as well. In that case I won't change the structure of it for now.

Reply:
We can use `MulAddChipset` to replicate this logic:
```rust
let left_shifters = Self::left_shifters();
let mut sum = N::ZERO;
for i in 0..NUM_LIMBS {
	sum = MulAddChipset::new(input[i], left_shifters[i], sum).synthesize(...);
}
sum
```
Where `left_shifters` are assigned to fixed columns.

For public keys, lets use the limbs for hashing. I think the collision attacks are possible if we compose them

Reply:
Okay, I got it and I added TODO for it and I will do it after the meeting.